client_code/historic/services.py

```diff
@@ -0,0 +1,93 @@
+from anvil_labs import non_blocking
+from anvil_labs.atomic import action, atom, reaction, selector
+
+__version__ = "0.0.1"
+
+
+@atom
+class Archivist:
+    def __init__(self, publisher, projectors=None, deferral=2):
+        """A class to sending events to the server for persistence.
+
+        Parameters
+        ----------
+        publisher : anvil_extras.messaging.Publisher
+        projectors : list
+            of projector names to run after events have been saved
+        deferral : int
+            number of seconds to wait before retrying if a save is already in progress
+        """
+        self.projectors = projectors or []
+        self.deferral = deferral
+        self.saving = []
+        self.pending = []
+        self.failed = []
+        publisher.subscribe(
+            channel="event", subscriber=self, handler=self._handle_message
+        )
+        reaction(lambda: self.pending, self._update_db)
+        self.deferred_save = None
+
+    @property
+    @selector
+    def status(self):
+        pending, saving, failed = self.pending, self.saving, self.failed
+        if failed:
+            return "unsaved"
+        if pending or saving:
+            return "saving"
+        return "saved"
+
+    def _handle_message(self, message):
+        if message.title == "historic.event.occurred":
+            self.pending.append(message.content)
+
+        if message.title == "historic.events.occurred":
+            self.pending.extend(message.content)
```

Comment:
duplicate code

Reply:
not quite. append vs extend

Reply:
The first handles messages containing a single event; the second, messages with a batch of events.